Mathlib/Analysis/Oscillation.lean

```diff
@@ -0,0 +1,125 @@
+at `x` is 0 if and only if `f` is continuous at `x`, with versions for both `oscillation` and
+`oscillationWithin`.
+
+## Tags
+
+oscillation, oscillationWithin
+-/
+
+open Topology EMetric
+
+universe u v
+
+variable {E : Type u} {F : Type v} [PseudoEMetricSpace F]
+
+/-- The oscillation of `f : E → F` at `x`. -/
+noncomputable def oscillation [TopologicalSpace E] (f : E → F) (x : E) : ENNReal :=
+  ⨅ S ∈ (𝓝 x).map f, diam S
+
+/-- The oscillation of `f : E → F` within `D` at `x`. -/
+noncomputable def oscillationWithin [TopologicalSpace E] (f : E → F) (D : Set E) (x : E) :
+  ENNReal := ⨅ S ∈ (𝓝[D] x).map f, diam S
+
+/-- The oscillation of `f` at `x` within a neighborhood `D` of `x` is equal to `oscillation f x` -/
+theorem oscillationWithin_nhd_eq_oscillation [TopologicalSpace E] (f : E → F) (x : E) (D : Set E)
+    (hD : D ∈ 𝓝 x) : oscillationWithin f D x = oscillation f x := by
+  rw [oscillation, oscillationWithin, nhdsWithin_eq_nhds.2 hD]
+
+/-- The oscillation of `f` at `x` within `Set.univ` is equal to `oscillation f x` -/
+theorem oscillationWithin_univ_eq_oscillation [TopologicalSpace E] (f : E → F) (x : E) :
+    oscillationWithin f Set.univ x = oscillation f x :=
+  oscillationWithin_nhd_eq_oscillation f x Set.univ Filter.univ_mem
+
+/-- The oscillation within `D` of `f` at `x ∈ D` is 0 if and only if `ContinuousWithinAt f D x`. -/
+theorem oscillationWithin_zero_iff_continuousWithinAt [TopologicalSpace E] (f : E → F) {D : Set E}
+    {x : E} (xD : x ∈ D): oscillationWithin f D x = 0 ↔ ContinuousWithinAt f D x := by
+  refine ⟨fun hf ↦ EMetric.tendsto_nhds.mpr (fun ε ε0 ↦ ?_), fun hf ↦ ?_⟩
+  · simp_rw [← hf, oscillationWithin, iInf_lt_iff] at ε0
+    obtain ⟨S, hS, Sε⟩ := ε0
+    refine Filter.mem_of_superset hS (fun y hy ↦ lt_of_le_of_lt ?_ Sε)
+    exact edist_le_diam_of_mem (Set.mem_preimage.1 hy) <|
+            Set.mem_preimage.1 (mem_of_mem_nhdsWithin xD hS)
+  · refine le_antisymm (ENNReal.le_of_forall_pos_le_add fun ε hε _ ↦ ?_) (zero_le _)
+    rw [zero_add]
+    have : ball (f x) (ε / 2) ∈ (𝓝[D] x).map f := hf <| ball_mem_nhds _ (by simp [ne_of_gt hε])
+    refine (biInf_le diam this).trans (le_of_le_of_eq diam_ball ?_)
+    exact (ENNReal.mul_div_cancel' (by norm_num) (by norm_num))
+
+/-- The oscillation of `f` at `x` is 0 if and only if `f` is continuous at `x`. -/
+theorem oscillation_zero_iff_continuousAt [TopologicalSpace E] (f : E → F) (x : E) :
+    oscillation f x = 0 ↔ ContinuousAt f x := by
+  rw [← oscillationWithin_univ_eq_oscillation, ← continuousWithinAt_univ f x]
+  exact oscillationWithin_zero_iff_continuousWithinAt f (Set.mem_univ x)
+
+/-- If `oscillationWithin f D x < ε` at every `x` in a compact set `K`, then there exists `δ > 0`
+such that the oscillation of `f` on `ball x δ ∩ D` is less than `ε` for every `x` in `K`. -/
+theorem uniform_oscillationWithin_of_compact [PseudoEMetricSpace E] {K : Set E} (comp : IsCompact K)
```

Comment:
This should be named `IsCompact.uniform_oscillationWithin` so that it can be used using generalized projection notation (aka dot notation).

Reply:
Done. I also added ContinuousWithinAt.oscillationWithin_eq_zero and ContinuousAt.oscillation_eq_zero so those can be accessed with dot notation also, and I tried to pick nicer names for a few other theorems in that file.